File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>> finalRes;
        int i=0,n=nums.size();
        
        for(int i=0;i<n-1;i++)
        {
            if(i!=0 && nums[i] == nums[i-1]) continue;
            int j=i+1,k=n-1;
            
            while(j<k) {
            int curSum = nums[i]+nums[j]+nums[k];
            if(curSum == 0) {
                vector<int> curRes = {nums[i],nums[j],nums[k]};
                finalRes.push_back(curRes);
                j++;
                k--;
                                while(j<k && nums[j] == nums[j-1]) j++;
                while(j<k && nums[k] == nums[k+1]) k--;

            }
            else if(curSum<0)
            {
                j++;
                while(j<k && nums[j] == nums[j-1]) j++;
            }
            else
            {
                k--;
                while(j<k && nums[k] == nums[k+1]) k--;
            }
            }
        }
        return finalRes;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        
        int target = 0;
        
        List<List<Integer>> result = new ArrayList<>();
        
        Arrays.sort(nums);
        
        for(int i=0;i<nums.length;i++)
        {
            if(i!=0 && nums[i]==nums[i-1]) continue;
            int j= i+1;
            int k = nums.length-1;        
            while(j<k)
            {
              

                if(nums[i]+nums[j]+nums[k] < target) 
                {
                    j++;   
                }   
                else if(nums[i]+nums[j]+nums[k] > target)
                {
                    k--;        
                }
                else 
                {
                    List<Integer> currRes = new ArrayList<>();
                    currRes.add(nums[i]);
                    currRes.add(nums[j]);
                    currRes.add(nums[k]);    
                    
                    result.add(currRes);
                    
                    int currK = nums[k];
                    while(nums[k] ==currK && j<k) k--;
                    int currJ = nums[j];
                    while(nums[j] == currJ && j<k) j++;
                }
            }
        }
        
        return result;
    }
}